supabase/migrations/soft_delete_flashcard_sets.sql

```diff
@@ -0,0 +1,59 @@
+-- Add is_deleted column to public.flashcard_sets
+ALTER TABLE
+  public.flashcard_sets
+ADD
+  COLUMN IF NOT EXISTS is_deleted BOOLEAN NOT NULL DEFAULT FALSE;
+
+-- Add deleted_at column to public.flashcard_sets
+ALTER TABLE
+  public.flashcard_sets
+ADD
+  COLUMN IF NOT EXISTS deleted_at timestamp WITH TIME ZONE;
+
+-- Add is_deleted column to public.flashcards
+ALTER TABLE
+  public.flashcards
+ADD
+  COLUMN IF NOT EXISTS is_deleted BOOLEAN NOT NULL DEFAULT FALSE;
+
+-- Add deleted_at column to public.flashcards
+ALTER TABLE
+  public.flashcards
+ADD
+  COLUMN IF NOT EXISTS deleted_at timestamp WITH TIME ZONE;
+
+-- Create or replace function to soft delete public.flashcard_sets and public.flashcards
+CREATE
+OR REPLACE FUNCTION flashcard_sets_soft_delete_fn() RETURNS TRIGGER AS $fn$
+  BEGIN
+    UPDATE public.flashcard_sets SET is_deleted=true, deleted_at=now() WHERE id=OLD.id;
+    UPDATE public.flashcards SET is_deleted=true, deleted_at=now() WHERE set_id=OLD.id;
+    RETURN NULL;
+  END;
+$fn$ LANGUAGE plpgsql;
+
+-- Drop trigger 'before delete public.flashcard_sets' before create
+DROP TRIGGER IF EXISTS flashcard_sets_soft_delete_trigger ON public.flashcard_sets CASCADE;
+
+-- Add trigger to before delete public.flashcard_sets
+CREATE TRIGGER flashcard_sets_soft_delete_trigger BEFORE DELETE ON public.flashcard_sets FOR EACH ROW EXECUTE FUNCTION flashcard_sets_soft_delete_fn();
+
+-- Create a view to hide deleted flashcard_sets
+CREATE
+OR REPLACE VIEW flashcard_sets_view AS
+SELECT
+  *
+FROM
+  public.flashcard_sets
+WHERE
+  is_deleted = false;
```

Comment:
We can use `where deleted_at is null`. Don't forget to create an index for `deleted_at`.